### DISCUSSION TITLE
AssertionError when running artisan

### DISCUSSION THREAD

**Discussion by italoaa** [firefly-iii/firefly-iii#9315] [Firefly III]
### Support guidelines

- [x] I've read the [support guidelines](https://github.com/firefly-iii/firefly-iii/blob/main/.github/support.md)

### I've found a bug and checked that ...

- [x] ... [the documentation](https://docs.firefly-iii.org/) does not mention anything about my problem
- [x] ... there are no open or closed issues that are related to my problem
- [x] ... it's [definitely a Firefly III issue, not me](https://github.com/firefly-iii/firefly-iii/blob/main/.github/its_you_not_me.md)

### Description

So I installed firefly-iii through nix. I have tried my best to debug the issue but I am completely incapable of doing so. I hope its a firefly III issue and not my stupidity. 

I am running firefly iii version 6.1.21. when ever I run `./artisan`:

```[2024-10-05 12:48:45] local.ERROR: Exception is: {"class":"AssertionError","errorMessage":"JSON:API server 'v2' does not exist in config or is not a valid class.","time":"Sat, 05 Oct 2024 12:48:45 +0000","file":"\/nix\/store\/hgwn5ddhxk7g1y7mbfjcsi1bxi3zzr1r-firefly-iii-6.1.21\/vendor\/laravel-json-api\/core\/src\/Core\/Server\/ServerRepository.php","line":76,"code":1,"version":"6.1.18","url":"http:\/\/192.168.0.30","userAgent":"Symfony","json":true,"method":"GET","post":""}
[2024-10-05 12:48:45] local.DEBUG: reachedLimit()
[2024-10-05 12:48:45] local.DEBUG: Read file in "/nix/store/hgwn5ddhxk7g1y7mbfjcsi1bxi3zzr1r-firefly-iii-6.1.21/storage/framework/cache/error-count.json"
[2024-10-05 12:48:45] local.DEBUG: Now checking limit "5m" {"limit":5,"reset":300}
[2024-10-05 12:48:45] local.DEBUG: Time past for this limit is 3002 seconds, exceeding 300 seconds. Reset to zero.
[2024-10-05 12:48:45] local.DEBUG: Now checking limit "1h" {"limit":15,"reset":3600}
[2024-10-05 12:48:45] local.DEBUG: Time past for this limit is 3631 seconds, exceeding 3600 seconds. Reset to zero.
[2024-10-05 12:48:45] local.DEBUG: Now checking limit "24h" {"limit":15,"reset":86400}
[2024-10-05 12:48:45] local.WARNING: Sent 16 emails in 24h, return true.
[2024-10-05 12:48:45] local.INFO: MailError: reached limit, not sending email.
[2024-10-05 12:48:45] local.ERROR: JSON:API server 'v2' does not exist in config or is not a valid class. {"exception":"[object] 

...

"}

   AssertionError

  JSON:API server 'v2' does not exist in config or is not a valid class.

  at vendor/laravel-json-api/core/src/Core/Server/ServerRepository.php:76
     72▕     private function make(string $name): ServerContract
     73▕     {
     74▕         $class = $this->classes[$name] ?? $this->config()->get("jsonapi.servers.{$name}");
     75▕
  ➜  76▕         assert(
     77▕             !empty($class) && class_exists($class) && is_a($class, ServerContract::class, true),
     78▕             "JSON:API server '{$name}' does not exist in config or is not a valid class.",
     79▕         );
     80▕

      +4 vendor frames

  5   routes/api.php:243
      Illuminate\Support\Facades\Facade::__callStatic("server")
      +4 vendor frames

  10  app/Providers/RouteServiceProvider.php:46
      Illuminate\Routing\RouteRegistrar::group("/nix/store/hgwn5ddhxk7g1y7mbfjcsi1bxi3zzr1r-firefly-iii-6.1.21/routes/api.php")
```

### Debug information

the server is not running I can get to it. so i cant include the debug

### Expected behaviour

I have a previous version installed in my system and that works fine the artisan runs perfectly.

### Steps to reproduce

with nix:
1. i have the following config:
```nix
  services.firefly-iii = {
    enable = true;
    user = "<user>";
    enableNginx = true;
    virtualHost = "http://192.168.0.30";
    settings.APP_KEY_FILE = "/etc/firefly-iii/firefly-iii.key";
    settings.APP_URL = "http://192.168.0.30";
    settings.APP_LOG_LEVEL = "debug";
    settings.MAIL_DRIVER = "log";
  };

```

2. On rebuild the setup service fails when calling the artisan.

### Additional info

I hope this info helps but the previous version i had would work perfectly so I would like to belive that the configuration in my part is okay, my reasoning is that if before it was perfect (v6.1.18) and then i run `nix flake update` to (v6.1.21) and it stopped working it means there is a problem in the package. but i am not familiar yet with the inner workings of ff3 so I could be in the wrong.

**Comment by JC5** [comment]
Hi, thanks for opening an issue. 

Why are you calling artisan in the first place?

**Comment by italoaa** [comment]
Hi, sorry for the delay. The issue is that the nix package sets up a systemd service that calls a script called "firefly-iii maintainance" and this script does the following:

```
       File: /nix/store/sjbdqww7cb63gz2ab5zcqj5famj9583h-firefly-iii-maintenance.sh
   1   #!/nix/store/8jind2ncd72hlzym9pghjjdj79fvqzv5-bash-5.2p32/bin/bash
   2   set -a
   3   APP_ENV=local
   4   APP_LOG_LEVEL=debug
   5   APP_URL=http://localhost
   6   DB_CONNECTION=sqlite
   7   DB_HOST=localhost
   8   DB_PORT=''
   9   MAIL_DRIVER=log
  10   APP_KEY="$(< /etc/firefly-iii/firefly-iii.key)"
  11
  12   set +a
  13
  14
  15   touch /var/lib/firefly-iii/storage/database/database.sqlite
  16   /nix/store/hgwn5ddhxk7g1y7mbfjcsi1bxi3zzr1r-firefly-iii-6.1.21/artisan cache:clear
  17   /nix/store/hgwn5ddhxk7g1y7mbfjcsi1bxi3zzr1r-firefly-iii-6.1.21/artisan package:discover
  18   /nix/store/hgwn5ddhxk7g1y7mbfjcsi1bxi3zzr1r-firefly-iii-6.1.21/artisan firefly-iii:upgrade-datab
       ase
  19   /nix/store/hgwn5ddhxk7g1y7mbfjcsi1bxi3zzr1r-firefly-iii-6.1.21/artisan firefly-iii:laravel-passp
       ort-keys
  20   /nix/store/hgwn5ddhxk7g1y7mbfjcsi1bxi3zzr1r-firefly-iii-6.1.21/artisan view:cache
  21   /nix/store/hgwn5ddhxk7g1y7mbfjcsi1bxi3zzr1r-firefly-iii-6.1.21/artisan route:cache
  22   /nix/store/hgwn5ddhxk7g1y7mbfjcsi1bxi3zzr1r-firefly-iii-6.1.21/artisan config:cache
```

By the above script I can tell that it crashes once artisan is being called from the journal ctl logs. After investigating further I went directly to artisan to do the calls myself but it caused the same "JSON:API server 'v2' does not exist in config or is not a valid class."

**Comment by JC5** [comment]
Mmm, I see what you mean. But I know very little about the nix package, I didn't create it.

Did the creator install all necessary dependencies? Are they included?

**Reply by savyajha** [reply]
Hi @JC5, the necessary dependencies are installed and the package works well. It lacks all the translations because we build from source using `composer` and `npm`, but everything apart from that is fine.

The reason there is an issue is because `artisan cache:clear` seems to not clear the cache. I'm not sure whether this is expected behaviour, but in any case, I'll change the maintenance script to do that manually instead.

**Reply by JC5** [reply]
Weird. The command comes from the parent framework. It should clear the cache indeed, but if it can be done manually that works too.

**Comment by italoaa** [comment]
I think so ... I will go throug the nix package later but thank you anyways :) i wanted to make sure the error above was not a easy fix. I will put any updates here though.

**Comment by timhae** [comment]
Hey @italoaa I faced the same issue, can you please go into your firefly-iii state dir (`/var/lib/firefly-iii/`) and remove the cache folder, then try to (re-)start the `firefly-iii-setup.service`? The other change I have made was to overwrite the artisan commands in the maintenance script to whatever is written in the `composer.json` post-update-cmd section. @JC5 should those be preferred over the ones given [here](https://docs.firefly-iii.org/how-to/firefly-iii/upgrade/self-managed/#run-upgrade-commands)? Anyways, sorry for the noise, this is a nixpkgs downstream packaging issue and will be resolved there..

**Reply by italoaa** [reply]
Thank you so much for input!! I was so lost weather it could be a ff3 issue or nix. Im glad to hear the issue is in the nix package. I will go over there to discuss. @JC5 sorry for the confusion

**Reply by JC5** [reply]
Not a problem!
